sound/hybrid_font.h

```diff
@@ -119,8 +119,12 @@ class HybridFont : public SaberBase {
     }
     hum_player_ = next_hum_player_;
     next_hum_player_.Free();
-    hum_player_->PlayOnce(f);
-    current_effect_length_ = hum_player_->length();
+
+    if (f) {
```

Comment:
I'm confused, what is this supposed to do?

Reply:
So THIS is what fixed the hum. Without a power sound it was trying to play a nullptr and throwing weird stuff in serial monitor. This just ensured it was not trying to play a non existent file and allowed hum to start properly. It is a file existence safety check to make sure a poweron or out exists before trying to play it because in blasters they don't

Reply:
Hmm, so when was it trying to play a null? During lockup handling? Or somewhere else?
I'm trying to make sure that this is the right fix and not just hiding a problem.

Reply:
SB_On was calling PlayMonophonic(NULL, &SFX_hum).

Because PlayMonophonic takes two effect arguments but only the second one had a check to see if it actually was there. So it assumed the first argument was always a properly formed Effect pointer and in the case of blaster it is not.

Reply:
SB_On() does not call PlayMonophonic(NULL, &SFX_hum);
It calls PlayMonophonic(&SFX_poweron, &SFX_hum);
Now, SFX_poweron might not have any files in it, but that's still different from a null pointer.
Importantly, the check you have here tests for null pointers, not pointers to effects without files, so I don't think this if statement does anything in SB_On(). It might do something in other places where PlayMonophonic is might actually be called with a null pointer, but that seems like it is not what you intended...

Reply:
Yes I know the call was not explicitly NULL. With this if check, hum starts correctly. Without it, serial monitor shows it trying to play a garbage file (looks like a memory address to something not correct) and hum doesn't start until a different effect is fired. So SOMETHING is not "just working" in an immediate call to On() with no poweron file.

Reply:
It's not enough that it works, we must also understand why.

The only way this actually does anything is if f is NULL. If f is not null, the code will behave the same as before the change. As far as I can tell, SB_On() does not call PlayMonophonic with f=NULL. The only places that may call PlayMonophonic() with f=NULL is SB_BeginLockup() and SB_EndLockup(), so if this code fixes the hum sound, then it must have something to do with the lockup functions.

Reply:
Ok. Let me back up and explain what I am seeing in the code.

hybrid_font. In Activate(). monophonic_hum_ is set by OR'ing a bunch of potential monophonic sounds: SFX_poweon, SFX_poweoff, etc. you added SFX_blast to this with the earlier font detection fix you made, which was good. However. In a blaster it is the only thing that returns true. So, monophonic_hum_ is now NOT null. 

Then we jump to SB_On() which checks monophonic_hum_ (it will pass this check and get inside the block). Inside that block it calls PlayMonophonic(&SFX_poweron, &SFX_hum). In blasters, the first argument is NOT assigned a proper Effect pointer.

Now. In PlayMonophonic, it just blindly calls Play() on the underlying player without any verification of the first argument. 

The underlying player gets Play() called on it without a properly formed Effect pointer. In fact, it could be any memory segment that has been cast to an Effect pointer because in this case nothing has actually assigned it or initialized it because poweron does not exist.

The bug here is that if a font is determined to be monophonic and monophonic_hum_ is assigned, then there is an unsafe and implicit assumption SFX_poweron is a properly formed Effect pointer.

The if (f) is trying to ensure that it is not a null. If it should not be an if (f) and should do something other than that, then I guess I fail to understand why things like if (SFX_<anything>) type of checks in the code elsewhere work. We could make it an if (f.files_found()) but i think the problem is not that the font is empty. It is the pointer had not been properly initialized because poweron is assumed to exist in this code path and it doesn't.

It has nothing to do with lockup.

Reply:
You're mixing up objects and pointers to objects, which is why there is some confusion here.
SFX_poweron is a global instance of the Effect class. It is a memory region, and it's existence or non-existence is not predicated by if any wav files are found or not. It always exists.

When we use SFX_poweron in an if statement, what's actually happening is that the function Effect::operator bool() gets called. It checks if any files were found and returns true or false based on that. The compiler will do this whenever we use any of the SFX_* objects as a boolean.

Because I don't want to make copies of Effect objects, all the methods that operate on Effects take a pointer to an Effect object as argument. (Like "Effect *effect")  Ampersand is used to get the pointer to the object, and that pointer will point to the object, regardless of whether it any wav files have been found or not.  When a pointer is used as a boolean, Effect::operator bool() is not called. Only null pointers are treated as false.

Reply:
Part of this is I am miscommunicating. Yes I know how pointers work and are dereferenced (it just has been almost 15 years since I wrote any significant amount of C/C++ code).

I also am aware that calling if on an object will use the bool() method on the class. And if on a pointer to the object will check if the pointer itself is null (null pointer).

So... then we need to understand why in the case of blaster is a pointer to SFX_poweron a null pointer when the poweron does not exist in the font.

How do I get to the Arduino equivalent of gdb? I want to start stepping through the code.

Reply:
I only know how to run gdb on proffieboards on linux.  (Not sure how to do it on teensys)
You also need an st-link v2 hooked up to the debug pins on the board, then you start up then you follow this post:

http://therebelarmory.com/post/113056/thread

Although, nowadays it seems better to use "gdb-multiarch" from the linux system instead of the gdb that comes with the arduino tools.

Reply:
Ugh. Ok so after hours of scouring the PJRC forums... looks like Teensy does not have a good way to do this in the 3.2 hardware. 3.5+ it added debug pins.

So... how do you suggest digging into this? Printf statements everywhere and serial monitor I guess?

Reply:
printf is probably the only thing you can do right away.
It's a reasonably good way to debug anyways, debuggers are really much better for inspecting the current state than for figuring out the flow of a program.

Reply:
So I tried this with and without the if statement and the hum worked equally well either way as far as I can tell.

Reply:
Alright. Well for now I removed this. Maybe the other changes have fixed this. I will retest everything and see if the hum still starts correctly but for the sake of finishing up this PR I removed the nullptr check since it was the last blocker.